ocaml/libs/tracing/tracing.ml

```diff
@@ -78,15 +78,91 @@ module SpanEvent = struct
 end
 
 module SpanContext = struct
-  type t = {trace_id: string; span_id: string} [@@deriving rpcty]
+  type t = {
+      trace_id: string
+    ; span_id: string
+    ; trace_flags: char
+    ; tracestate: (string * string) list
+    ; is_remote: bool
+  }
+
+  let char_to_octet c =
+    Hex.of_char c |> fun (a, b) -> String.of_seq Seq.(cons a (return b))
 
-  let to_traceparent t = Printf.sprintf "00-%s-%s-00" t.trace_id t.span_id
+  let char_of_octet f =
+    try "0x" ^ f |> int_of_string |> char_of_int with _ -> '\x00'
 
-  let of_traceparent traceparent =
+  let is_valid t =
+    String.exists (( <> ) '0') t.span_id
+    && String.exists (( <> ) '0') t.trace_id
+
+  let generate_id n = String.init n (fun _ -> "0123456789abcdef".[Random.int 16])
```

Comment:
This seems a pretty inefficient way of generating the ids

This function is used to generate identifiers of 64 bits (16 hexadecimal digits) and 128 bits (32 hexadecimal digits)
I believe using one or two calls to [Random.bits64](https://v2.ocaml.org/api/Random.html#VALbits64) and serializing with hex might be faster. Tee serialization can be done right away or delay it until it's needed, depending on how often they are serialized it might be worth it, memory and time-wise

Reply:
I once proposed this. You get one call to `Random.int` per character (4bit) - so indeed this could be more efficient it we need this a lot.